internal/services/compute/virtual_machine_implicit_data_disk_resource.go

```diff
@@ -0,0 +1,462 @@
+			}
+
+			schema.Lun = disk.Lun
+			schema.Caching = string(pointer.From(disk.Caching))
+			schema.CreateOption = string(disk.CreateOption)
+			schema.DeleteOption = string(pointer.From(disk.DeleteOption))
+			schema.DiskSizeGb = pointer.From(disk.DiskSizeGB)
+			if disk.SourceResource != nil {
+				schema.SourceResourceId = pointer.From(disk.SourceResource.Id)
+			}
+
+			schema.WriteAcceleratorEnabled = pointer.From(disk.WriteAcceleratorEnabled)
+
+			return metadata.Encode(&schema)
+		},
+	}
+}
+
+func (r VirtualMachineImplicitDataDiskResource) Delete() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Compute.VirtualMachinesClient
+
+			id, err := parse.DataDiskID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			locks.ByName(id.VirtualMachineName, VirtualMachineResourceName)
+			defer locks.UnlockByName(id.VirtualMachineName, VirtualMachineResourceName)
+
+			virtualMachineId := virtualmachines.NewVirtualMachineID(id.SubscriptionId, id.ResourceGroup, id.VirtualMachineName)
+			resp, err := client.Get(ctx, virtualMachineId, virtualmachines.DefaultGetOperationOptions())
+			if err != nil {
+				return fmt.Errorf("retrieving %s: %+v", virtualMachineId, err)
+			}
+
+			if model := resp.Model; model != nil {
+				if props := model.Properties; props != nil {
+					if profile := props.StorageProfile; profile != nil {
+						newDisks := make([]virtualmachines.DataDisk, 0)
+						var toBeDeletedDisk *virtualmachines.DataDisk
+						if dataDisks := profile.DataDisks; dataDisks != nil {
+							for _, dataDisk := range *dataDisks {
+								if pointer.From(dataDisk.Name) != id.Name {
+									newDisks = append(newDisks, dataDisk)
+								} else {
+									toBeDeletedDisk = &dataDisk
+								}
+							}
+						}
+
+						profile.DataDisks = &newDisks
+
+						// fixes #24145
+						model.Properties.ApplicationProfile = nil
+
+						// fixes #2485
+						model.Identity = nil
+						// fixes #1600
+						model.Resources = nil
+
+						err = client.CreateOrUpdateThenPoll(ctx, virtualMachineId, *model, virtualmachines.DefaultCreateOrUpdateOperationOptions())
+						if err != nil {
+							return fmt.Errorf("deleting %s: %+v", id, err)
+						}
+
+						// delete data disk if delete_option is set to delete
+						if toBeDeletedDisk != nil && pointer.From(toBeDeletedDisk.DeleteOption) == virtualmachines.DiskDeleteOptionTypesDelete &&
+							toBeDeletedDisk.ManagedDisk != nil && toBeDeletedDisk.ManagedDisk.Id != nil {
+							diskClient := metadata.Client.Compute.DisksClient
+							diskId, err := commonids.ParseManagedDiskID(*toBeDeletedDisk.ManagedDisk.Id)
+							if err != nil {
+								return err
+							}
+
+							err = diskClient.DeleteThenPoll(ctx, *diskId)
+							if err != nil {
+								return fmt.Errorf("deleting Managed Disk %s: %+v", *diskId, err)
+							}
+						}
```

Comment:
This is an API setting - why are we deleting the disk here?

Reply:
Because this disk is created by the backend so that it's not managed by any TF resource such as `azurerm_managed_disk`. If we do not delete it explicitly here, it will only be removed from the VM configuration, but the disk itself still exists. When VM is deleted, this disk will not be deleted together because it's already been removed from the VM configuration. Since the disk still exists in the resource group, the resource group cannot be deleted unless `prevent_deletion_if_contains_resources=false`. Due to this reason, I choose to explicitly delete the disk when `DeleteOption=true`

Reply:
Hmmm @ms-zhenhua, what're your thoughts on deleting the disk when it's unhooked from the vm and not worrying about `delete_option`? This resource means that Terraform is creating the disk and I would expect Terraform to also delete the disk

Are there use cases when we want to keep the disk around?

Reply:
Hi @mbfrahry, when we create this resource, what we really do is to add a data disk configuration to the existing VM, which will trigger the Azure backend to create an implicit data disk for the VM. However, when we delete this resource, what we really do is just remove the data disk configuration from the VM, but the physical data disk is only detached from the VM and still exists in the resource group. That's why I use the delete option to manually delete this data disk, otherwise, the resource group cannot be deleted normally.